Consistently support storing multiple subresource frames when storage access is granted

sysrqb opened the pull request:
#### f93c856c879e9f46a2d8e2de3cdbc671769069b2
<pre>
Consistently support storing multiple subresource frames when storage access is granted
<a href="https://bugs.webkit.org/show_bug.cgi?id=269965">https://bugs.webkit.org/show_bug.cgi?id=269965</a>
<a href="https://rdar.apple.com/123485237">rdar://123485237</a>

Reviewed by Sihui Liu.

Currently, some parts of ResourceLoadStatistics supports multiple subframe
domains for storing granted storage access, but others only support storing a
single domain. This change improves the consistency by storing a list of
subframe domains in a vector. I don&apos;t expect that the set of subframe domains
should be large, so iterating over a vector is probably more efficient than
using a HashSet.

Adds a new test that covers behavior.

* Source/WebCore/loader/ResourceLoadObserver.h:
(WebCore::ResourceLoadObserver::setDomainsWithCrossPageStorageAccess):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setDomainsWithCrossPageStorageAccess):
(WebCore::NetworkStorageSession::hasStorageAccess const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::domainsWithStorageAccess const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::callUpdatePrevalentDomainsToBlockCookiesForHandler):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setShouldDirectlyCallRequestStorageAccessConfirmDelegateForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
(WebKit::WebResourceLoadObserver::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addDomainWithPageLevelStorageAccess):
(WebKit::WebPage::hasPageLevelStorageAccess const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.mm:
(-[TestUIDelegate _webView:requestStorageAccessPanelForDomain:underCurrentDomain:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/275890@main">https://commits.webkit.org/275890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c3ce996ebf6eb9a6aa615aea97d8ffc88ee752e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43057 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22083 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45459 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35646 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19119 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37082 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16644 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16710 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38139 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1113 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39240 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47232 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17959 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14756 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42398 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19509 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41051 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9607 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19688 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -230,11 +230,13 @@ void NetworkStorageSession::setDomainsWithUserInteractionAsFirstParty(const Vect
     m_pairsGrantedCrossPageStorageAccess.clear();
-    for (auto& topFrameDomain : domains.keys())
-        grantCrossPageStorageAccess(topFrameDomain, domains.get(topFrameDomain));
+    for (auto topFrameDomainIt : domains) {
```
`for (auto& [topDomain, subDomains] : domains)`

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -269,7 +271,7 @@ bool NetworkStorageSession::hasStorageAccess(const RegistrableDomain& resourceDo
 
         auto it = m_pairsGrantedCrossPageStorageAccess.find(firstPartyDomain);
-        if (it != m_pairsGrantedCrossPageStorageAccess.end() && it->value.contains(resourceDomain))
+        if (it != m_pairsGrantedCrossPageStorageAccess.end() && it->value.isValidValue(resourceDomain) && it->value.contains(resourceDomain))
```
Why do we need to validate resourceDomain? When is it invalid?

sysrqb replied:
I suppose it might be safe by the time we reach this point.

szewai commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -437,7 +437,7 @@
 }
 
-void UIDelegate::UIClient::requestStorageAccessConfirm(WebPageProxy& webPageProxy, WebFrameProxy*, const WebCore::RegistrableDomain& requestingDomain, const WebCore::RegistrableDomain& currentDomain, std::optional<WebCore::OrganizationStorageAccessPromptQuirk>&& organizationStorageAccessPromptQuirk, CompletionHandler<void(bool)>&& completionHandler)
+void UIDelegate::UIClient::requestStorageAccessConfirm(WebPageProxy& webPageProxy, WebFrameProxy*, const WebCore::RegistrableDomain& requestingDomain, const WebCore::RegistrableDomain& currentDomain, std::optional<WebCore::OrganizationStorageAccessPromptQuirk>&& organizationStorageAccessPromptQuirk, bool shouldDirectlyCallDelegate, CompletionHandler<void(bool)>&& completionHandler)
```
It's unnecessary to add shouldDirectlyCallDelegate to this function because it can be retrieved from WebPageProxy. 
Also, why do we need to skip the other checks?

sysrqb replied:
Ah, good point about accessing it on `webPageProxy`. I want this for using in API tests so I can implement a delegate and I don't want to use the special-cases for storage access quirks.

szewai replied:
But we could craft the test to make it bypass the quirks check right? E.g. 127.0.0.1 should not be in quirk list.
Other than the quirk checks, the other check is `webViewRequestStorageAccessPanelUnderFirstPartyCompletionHandler`, which can be skipped by not implementing the method in the delegate.
I think we should avoid introducing testing-only SPI if they are not necessary. It usually makes the code more complicated and may not test the real behavior that we want to cover.

sysrqb replied:
Yeah, the problem I have is that I'm setting a storage access quirk so `organizationStorageAccessPromptQuirk` is input into that function. Ideally we have a way to disable the quirk checks - is that a better path?

szewai replied:
I am not sure I understand this... Will a normal WebKit client's delegate method `requestStorageAccessPanelForDomain`  ever be invoked when `organizationStorageAccessPromptQuirk` is not null? If not, why do we test this?

If the test has to set `organizationStorageAccessPromptQuirk` and it needs to catch `requestStorageAccessConfirm` call, then we should probably add a new delegate method for that case and check that method before `presentStorageAccessAlertSSOQuirk`.

sysrqb replied:
> Will a normal WebKit client's delegate method requestStorageAccessPanelForDomain ever be invoked when organizationStorageAccessPromptQuirk is not null? If not, why do we test this?

No, at this point I don't expect a WebKit client will need this behavior. I need it for the testing scenario because I want to control granting storage access without actually showing the prompt (showing the actual prompt is useless in the API test).

> If the test has to set organizationStorageAccessPromptQuirk and it needs to catch requestStorageAccessConfirm call, then we should probably add a new delegate method for that case and check that method before presentStorageAccessAlertSSOQuirk.

Okay, I'll use this approach. Thanks

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2645,7 +2645,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 #endif
 
-    HashMap<WebCore::RegistrableDomain, WebCore::RegistrableDomain> m_domainsWithPageLevelStorageAccess;
+    HashMap<WebCore::RegistrableDomain, Vector<WebCore::RegistrableDomain>> m_domainsWithPageLevelStorageAccess;
```
We probably want to use `HashSet` for data members for deduplication.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
```diff
@@ -447,18 +447,22 @@ bool WebResourceLoadObserver::hasCrossPageStorageAccess(const SubFrameDomain& su
-
-        // Some sites have quirks where multiple login domains require storage access.
-        if (auto additionalLoginDomain = WebCore::NetworkStorageSession::findAdditionalLoginDomain(topDomain, domains.get(topDomain))) {
+        for (auto& subResourceDomain : domains.get(topDomain)) {
```
```
for (auto& [topDomain, subResourceDomains] : domains) {
    for (auto& subResourceDomain : subResourceDomains) {
    }
}
```
to reduce lookup.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
```diff
@@ -447,18 +447,22 @@ bool WebResourceLoadObserver::hasCrossPageStorageAccess(const SubFrameDomain& su
+        }
+
+        // Some sites have quirks where multiple login domains require storage access.
+        for (auto& subResourceDomain : domains.get(topDomain)) {
```
Why a different for loop here?

sysrqb replied:
I don't remember :( seems strange.

szewai commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -448,6 +449,17 @@
         return;
     }
 
+    if (m_uiDelegate->m_delegateMethods.webViewRequestStorageAccessPanelWithoutQuirksUnderCurrentDomainCompletionHandler) {
```
I think the new delegate method should be only handling quirk cases (instead of skipping the quirk check), so that when client specifies both methods, one of the delegate methods will be invoked (in the existing approach, only `webViewRequestStorageAccessPanelWithoutQuirksUnderCurrentDomainCompletionHandler` will be invoked).
Something like:
```
if (organizationStorageAccessPromptQuirk || additionalLoginDomain) {
    if (m_uiDelegate->m_delegateMethods.webViewRequestStorageAccessPanelFromQuirksUnderCurrentDomainCompletionHandler) {
         // ...
    }
}
```
And our test will be able to confirm storage access request is from quirks by providing both delegate methods.

sysrqb replied:
Okay, sounds good! I'll make that change.

szewai commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -1327,11 +1327,13 @@ void WebResourceLoadStatisticsStore::callUpdatePrevalentDomainsToBlockCookiesFor
-            auto requestingDomain = domainsToBlock.domainsWithStorageAccess.get(firstPartyDomain);
-            if (NetworkStorageSession::loginDomainMatchesRequestingDomain(firstPartyDomain, requestingDomain))
-                domainsWithStorageAccessQuirk.add(firstPartyDomain, requestingDomain);
+            auto requestingDomains = domainsToBlock.domainsWithStorageAccess.get(firstPartyDomain);
```
```
for (auto& [firsrtPartyDomain, requestingDomains] : domainsToBlock.domainsWithStorageAccess) {
    for (auto& requestingDomain : requestingDomains)
}
```
to avoid the lookup

szewai commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2197,11 +2197,13 @@ void WebProcess::setDomainsWithUserInteraction(HashSet<WebCore::RegistrableDomai
-        for (auto& webPage : m_pageMap.values())
-            webPage->addDomainWithPageLevelStorageAccess(domain, domains.get(domain));
+        for (auto& webPage : m_pageMap.values()) {
+            for (auto& subResourceDomain : domains.get(domain))
```
```
for (auto& [domain, subResourceDomains] : domains) {
    for (auto& subResourceDomain : subResourceDomains) {
    }
}
```

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+
+    bool didSetSite2CookieHeader { false };
+    bool didSendSite2CookieHeader { false };
+    HTTPServer httpServer2 { HTTPServer::UseCoroutines::Yes, [&](Connection connection) -> Task {
```
Why httpserver2 (there seems no server1)?

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+    bool didSetSite2CookieHeader { false };
+    bool didSendSite2CookieHeader { false };
+    HTTPServer httpServer2 { HTTPServer::UseCoroutines::Yes, [&](Connection connection) -> Task {
+        while (1) {
```
`while (true)`

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+        (NSString *)kCFStreamPropertyHTTPProxyPort: @(httpServer2.port()),
+    }];
+
+    auto configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
```
What's this setting for? Why can't we use` auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);`?

sysrqb replied:
Not needed anymore (I was setting a user gesture in a previous version).

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+    [dataStore _logUserInteraction:[NSURL URLWithString:@"http://site2.example/"] completionHandler:^{
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
```
It seems this does not wait because `done` is not set back to false.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+    __block bool gotRequestStorageAccessPanelForDomain { false };
+    auto uiDelegate = adoptNS([TestUIDelegate new]);
+    uiDelegate.get().requestStorageAccessPanelWithoutQuirksForDomain = ^(WKWebView *, NSString *, NSString *, void  (^completionHandler)(BOOL)) {
+        gotRequestStorageAccessPanelForDomain = true;
```
We may need to check the requesting domains here.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+    done = false;
+    EXPECT_WK_STREQ([uiDelegate waitForAlert], @"true");
+
+    [webView evaluateJavaScript:@"let iframeSite4 = document.createElement(\"iframe\"); iframeSite3.src = \"http://site4.example/page1\"; document.body.appendChild(iframeSite4); true" completionHandler:^(id value, NSError *error) {
```
I think you want to set location for iframeSite4 not iframeSite3?

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+
+            if (path == "http://site2.example/page1"_s || path == "http://site3.example/page1"_s || path == "http://site4.example/page1"_s) {
+                if (host == "site2.example"_s && headerFromRequest(request, "Cookie: "_s) == "exists=1"_s)
+                    didSendSite2CookieHeader = true;
```
If we are going to check cookies existence, we might also check cookies for site3 and site4 to ensure the behavior is expected. i.e. site4 request should not contain cookie.
I don't think we need to verify whether cookies are included in request though; there should already be tests for ensuring StorageAccess is consistent with cookies sending.

sysrqb replied:
The new behavior that I'm testing here is sending cookies in the initial request for the iframe document.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
+            auto host = headerFromRequest(request, "Host: "_s);
+            auto path = HTTPServer::parsePath(request);
+
+            if (host == "site2.example"_s && path == "http://site2.example/page4"_s) {
```
Instead of page1, page2, etc, let's rename them to something that is meaningful, e.g. main.html / iframe.html, or cookie-set.html / cookie-get.html, etc; so we could perform corresponding checks.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1731,3 +1737,180 @@ HTTPServer httpServer({
     TestWebKitAPI::Util::run(&done);
 }
+
+TEST(ResourceLoadStatistics, GrantMultipleSubFrameDomainsStorageAccess)
```
Based on my understanding for this change, what we want to test is only one delegate callback for granting storage access to multiple domains, so this test might be as simple as:
1. Set quirked domains
2. Load different sites in the main frame
3. Load an iframe from quirked domain, which invokes `requestStorageAccess` => we check the new delegate method is called and this returns true
4. Load another iframe from another quirked domain, which also invokes `requestStorageAccess` => we check this returns true and new delegate method is not called (or it could invoke `hasStorageAccess` and verifies it's true)

If we want to ensure non-quirked domain will go another path:
5. Load another frame from non-quirked domain, which invokes `hasStorageAccess` => we check the this returns false

sysrqb replied:
This test is a reduced test case for an issue I was seeing. I would like to keep this test, but I can add a separate test like you describe that is more focused on this change in behavior.

szewai commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -448,6 +449,22 @@
+    auto additionalLoginDomain = WebCore::NetworkStorageSession::findAdditionalLoginDomain(currentDomain, requestingDomain);
+
+    if (organizationStorageAccessPromptQuirk || additionalLoginDomain) {
+        if (m_uiDelegate->m_delegateMethods.webViewRequestStorageAccessPanelWithoutQuirksUnderCurrentDomainCompletionHandler) {
```
Technically this method is invoked when there are quirk, so we probably want to name it 
`_webView:reuqestStorageAccessPanelFromQuirksForDomains:underCurrentDomain:completionHandler:`, and we should include the quirked domains so client (TestRunnner) could verify the domains are expected.

szewai reviewed (commented):
Let's simplify the test a bit and make sure it can run as expected.

szewai reviewed (approved):
r=me

sysrqb committed:
Consistently support storing multiple subresource frames when storage access is granted
https://bugs.webkit.org/show_bug.cgi?id=269965
rdar://123485237

Reviewed by Sihui Liu.

Currently, some parts of ResourceLoadStatistics supports multiple subframe
domains for storing granted storage access, but others only support storing a
single domain. This change improves the consistency by storing a list of
subframe domains in a vector. I don't expect that the set of subframe domains
should be large, so iterating over a vector is probably more efficient than
using a HashSet.

Adds a new test that covers behavior.

* Source/WebCore/loader/ResourceLoadObserver.h:
(WebCore::ResourceLoadObserver::setDomainsWithCrossPageStorageAccess):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setDomainsWithCrossPageStorageAccess):
(WebCore::NetworkStorageSession::hasStorageAccess const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::domainsWithStorageAccess const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::callUpdatePrevalentDomainsToBlockCookiesForHandler):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setShouldDirectlyCallRequestStorageAccessConfirmDelegateForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
(WebKit::WebResourceLoadObserver::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addDomainWithPageLevelStorageAccess):
(WebKit::WebPage::hasPageLevelStorageAccess const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setDomainsWithCrossPageStorageAccess):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.mm:
(-[TestUIDelegate _webView:requestStorageAccessPanelForDomain:underCurrentDomain:completionHandler:]):

Canonical link: https://commits.webkit.org/275890@main
